plugins/calibre.koplugin/parser.lua

```diff
@@ -1,5 +1,6 @@
 -- parse "metadata.calibre" files
 local lj = require("lunajson")
+local rapidjson = require("rapidjson")
```

Comment:
I don't think we need changes on this file. We use it to produce a lua table. The table should be valid as is, and the checks to make sure it can be properly encoded back to json can be placed in Calibre:Metadata

Reply:
That would incur doing double-duty for the other parsers, which already handle it during parsing.

This felt more logical to do it this way here too ;).

Reply:
But there're no other parsers :)

I mean, we use different code paths for decoding json but just rapidjson for encoding it. And in this plugin we only encode json on wireless clients.

Wireless clients used to encode properly when dumping, except for empty arrays, as the bug in OP tell us. As long  as we could keep a single function taking care of encoding I think we would be better using plain tables for our own representation of the data.

Reply:
AFAICT, If you set JSON parsing to the safe parser, you *can* get it to parse stuff later dumped by the wireless client, which would lead to malformed data if we don't fix the parser.

Reply:
I'm not sure I entirely get your concerns. The rapidjson flag is entirely transparent in front, as we don't actually use metatables ourselves on those objects. So, yes, they're not *technically* plain tables anymore, but the point is entirely moot for us.

While I can *sort of* get the semantic reasoning behind the other two comments, I still think it's easier to just flag everything, and avoid any potential mishaps will the least amount of code. (And it's definitely more important *here*, compared to the init/error codepaths).

Reply:
Sorry for the delay :p 

I'm not entirely sure If I have a point anymore. But **if** I have one it should be something like:

```lua
local function slim(book, is_search)
    local slim_book = rapidjson.object({})
    for _, k in ipairs(is_search and search_used_metadata or used_metadata) do
        if k == "series" or k == "series_index" then
            slim_book[k] = book[k] or rapidjson.null
        elseif k == "tags" or k == "authors" then
            slim_book[k] = book[k] or rapidjson.array({})
        else
            slim_book[k] = book[k]
        end
```

is enough.

We always  call `slim` before `dump`. It happens on wireless transfers, at document addition/deletion.
If each book contains the full representation of its json object when dumping `rapidjson.dump(rapidjson.array(books)..` should produce the same output.


> you can get it to parse stuff later dumped by the wireless client, which would lead to malformed data if we don't fix the parser.

If we don't fix the `slim` function, that's always called before encoding to json :) but you just did, above :)

> I'm not sure I entirely get your concerns.

`parser.lua` is a json decoder, it shouldn't need to know how to encode json.

Reply:
> parser.lua is a json decoder, it shouldn't need to know how to encode json.

Okay, that's indeed what I sort of thought was going on ;).

I get the idea, my stance is simply that I'd rather have all the parsers behave the same way, since it's trivial to do so, and I see another benefit to that: because otherwise we would effectively have to deal with it in a post-processing stage (i.e., slim), where it'd be redundant for the other two.

Which leads me to...

> [slim] is enough.

Not quite. We'd also have to mark every individual object/array as such, even when they exist in the data from the parsers.

Granted, that would only boil down to

```diff
-            slim_book[k] = book[k] or rapidjson.array({})
+           slim_book[k] = rapidjson.array(book[k]) or rapidjson.array({})
```

But, still, I find that rather less elegant than assuring that the parsers took care of it as they went ;).

Reply:
allright!